File: spring-integration-poc/src/main/java/org/example/springintegrationpoc/controller/MessageController.java
package org.example.springintegrationpoc.controller;

import org.example.springintegrationpoc.gateway.MessageGateway;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/messages")
public class MessageController {

    @Autowired
    private MessageGateway messageGateway;

    @PostMapping("/transform")
    public String sendTransformMessage(@RequestBody String message) {
        messageGateway.sendToTransformChannel(message);
        return "Message sent to transform channel: " + message;
    }

    @PostMapping("/enrich")
    public String sendEnrichMessage(@RequestBody String message, @RequestParam String messageType) {
        messageGateway.sendToEnrichChannel(message, messageType);
        return "Message sent to enrich channel: " + message + " with type: " + messageType;
    }

    @PostMapping("/json")
    public String sendJsonTransformMessage(@RequestBody String message) {
        messageGateway.sendToJsonTransformChannel(message);
        return "Message sent to JSON transform channel: " + message;
    }

    @PostMapping("/direct")
    public String sendDirectMessage(@RequestBody String message, @RequestParam String messageType) {
        messageGateway.sendDirectMessage(message, messageType);
        return "Message sent directly to input channel: " + message + " with type: " + messageType;
    }

    @GetMapping("/health")
    public String health() {
        return "Spring Integration POC is running!";
    }
}

File: kotlin-coroutines-poc/src/main/kotlin/org/example/coroutinespoc/controller/MessageController.kt
package org.example.coroutinespoc.controller

import kotlinx.coroutines.delay
import org.example.coroutinespoc.model.Message
import org.springframework.web.bind.annotation.*
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

@RestController
@RequestMapping("/api/messages")
class MessageController {

    private val messages = ConcurrentHashMap<String, Message>()

    @GetMapping
    suspend fun getMessages(): List<Message> {
        delay(500)
        return messages.values.toList()
    }

    @GetMapping("/{id}")
    suspend fun getMessage(@PathVariable id: String): Message {
        delay(200)
        return messages[id] ?: throw NoSuchElementException("Message not found: $id")
    }

    @PostMapping
    suspend fun createMessage(@RequestBody message: Message): Message {
        val saved = message.copy(id = UUID.randomUUID().toString())
        messages[saved.id] = saved
        return saved
    }
}
